rpc: doc: Added `longpollid` and `data` params to `template_request`

ItIsOHM opened the pull request:
This PR will add the optional parameters `longpollid` and `data` to `template_request` as they were missing when calling `help getblocktemplate` in RPCHelpMan.

I request the maintainers to review this and let me know about any mistakes in the descriptions of the parameters.

This PR refers to the issue #27998

ItIsOHM committed:
Added `longpollid` and `data` params to `template_request` #27998

Added `longpollid` and `data` params to `template_request` in `getblocktemplate` #27998

ItIsOHM commented:
> This doesn't compile. Ideally, you compile and test any changes locally before opening a pull request.

I see, I'll try to compile it on my end and will let you know if there's any issue. Also, can you please check the description for the new parameters added if they need any changes?

luke-jr commented on src/rpc/mining.cpp:
```diff
@@ -566,6 +566,8 @@ static RPCHelpMan getblocktemplate()
                     {"segwit", RPCArg::Type::STR, RPCArg::Optional::NO, "(literal) indicates client side segwit support"},
                     {"str", RPCArg::Type::STR, RPCArg::Optional::OMITTED, "other client side supported softfork deployment"},
                 }},
+                {"longpollid", RPCArg::Type::STR, RPCArg::Optional::OMITTED, " "longpollid" of job to monitor for expiration; required and valid only for long poll requests"},
```
```suggestion
                {"longpollid", RPCArg::Type::STR, RPCArg::Optional::OMITTED, "Delay processing request until the result would vary significantly from the \"longpollid\" of a prior template."},
```

It doesn't necessarily imply the previous template is expired.

luke-jr commented on src/rpc/mining.cpp:
```diff
@@ -566,6 +566,8 @@ static RPCHelpMan getblocktemplate()
                     {"str", RPCArg::Type::STR, RPCArg::Optional::OMITTED, "other client side supported softfork deployment"},
                 }},
+                {"longpollid", RPCArg::Type::STR, RPCArg::Optional::OMITTED, " "longpollid" of job to monitor for expiration; required and valid only for long poll requests"},
+                {"data", RPCResult::Type::STR_HEX, RPCArg::Optional::OMITTED, "Block data encoded in hexadecimal"},
```
```suggestion
                {"data", RPCResult::Type::STR_HEX, RPCArg::Optional::OMITTED, "Proposed block data to check, encoded in hexadecimal; valid only for mode=\"proposal\""},
```

ItIsOHM commented:
> This doesn't compile. Ideally, you compile and test any changes locally before opening a pull request.

Hey so I was trying to build it after adding the changes you suggested, but i'm getting this error in VS, as I'm building on Windows 11. Can you please help what this means? 
![image](https://github.com/bitcoin/bitcoin/assets/62976081/7b6d4dc0-7fda-41fe-9c8d-77f1e6ad75a8)

 I followed [this](https://github.com/bitcoin/bitcoin/blob/master/build_msvc/README.md) guide and opened `bitcoin.sln` in VS.

ItIsOHM commented:
> I strongly recommend not building in Windows/MSVC, as it's a venture on its own to get that to work. In a Linux VM, or WSL2, it should be fairly straightforward.

Ohh I see! Then I'll install WSL2 and try building it from there. I suppose [this](https://github.com/bitcoin/bitcoin/blob/master/doc/build-windows.md) guide should be enough to get it working?

ItIsOHM commented:
Hey there, I fixed the descriptions as suggested, and ran the tests and was able to build Bitcoin Core locally successfully.

ItIsOHM commented on vcpkg:
Oh I'm sorry I must have forgotten to discard everything other than `mining.cpp`. Should I revert this commit, or something else?

ItIsOHM commented:
Hey @MarcoFalke should I revert the last commit? All checks except the lint test are passing.

ItIsOHM commented:
> You need to squash the commits. See https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

I think I did squash them, atleast that's what it shows in the commits tab.

ItIsOHM commented:
> Apologies, you did indeed!

No worries, thanks for the guidance throughout this PR!

ItIsOHM commented:
> Apologies, you did indeed! The commit title is "Deleted stray file `vcpkg`" now, though.

Should I amend this commit or leave it as it is?

ItIsOHM commented:
Hey there, is this PR ready to merge if there aren't any errors?

russeree commented on src/rpc/mining.cpp:
```diff
@@ -566,6 +566,8 @@ static RPCHelpMan getblocktemplate()
                     {"segwit", RPCArg::Type::STR, RPCArg::Optional::NO, "(literal) indicates client side segwit support"},
                     {"str", RPCArg::Type::STR, RPCArg::Optional::OMITTED, "other client side supported softfork deployment"},
                 }},
+                {"longpollid", RPCArg::Type::STR, RPCArg::Optional::OMITTED, "Delay processing request until the result would vary significantly from the \"longpollid\" of a prior template."},
```
1. This should not have a period '.' at the end. Since it is inconsistent with the rest of the documentation. 
2. Should start with a lowercase letter since only grammatical articles have capitalized first letters for this documentation.  (A, The, This)
3. The language used in BIP22 is a more concise definition ```"longpollid" of job to monitor for expiration; required and valid only for long poll requests```

portlandhodl reviewed (changes requested):
Some of my thoughts about this PR. Mostly regarding the grammatical consistency of these messages.

ItIsOHM commented on src/rpc/mining.cpp:
```diff
@@ -566,6 +566,8 @@ static RPCHelpMan getblocktemplate()
                     {"segwit", RPCArg::Type::STR, RPCArg::Optional::NO, "(literal) indicates client side segwit support"},
                     {"str", RPCArg::Type::STR, RPCArg::Optional::OMITTED, "other client side supported softfork deployment"},
                 }},
+                {"longpollid", RPCArg::Type::STR, RPCArg::Optional::OMITTED, "Delay processing request until the result would vary significantly from the \"longpollid\" of a prior template."},
```
Oh yes, I agree the descriptions should have the same grammatical style as the rest. I can make these changes for sure. Thanks @russeree for the help :D

russeree replied:
No problem, this is what I got running your code.

![image](https://github.com/bitcoin/bitcoin/assets/3104223/825b14e7-ae1b-4a97-8193-9a76a21bdd66)

stickies-v replied:
Are you still going to make these changes?

ItIsOHM replied:
Actually I wasn't sure if I should since I had removed the WIP tag and asked for a review, but I can change them right now if that's okay.

ItIsOHM replied:
Ahh okay, I'm sorry about that. Will make these changes asap and push another commit.

ItIsOHM commented:
Hello, just wondering if there is something left from my side, or is this ready for review and merging?

russeree commented:
> 

Bitcoin development is always bottlenecked and all changes no matter how small must thoroughly reviewed for and allowed time for social consensus since every change could potentially cost people real world value. You can work on other issues and/or await more feedback.

ItIsOHM commented:
Ahh I see! Thanks for clearing it up :D I was just worried if I needed to change something from my side hehe

Thanks for all the help in this PR everyone, really learnt a great deal from each one of you!

achow101 commented:
If this is ready for review, please remove "[WIP]" from the title.

stickies-v reviewed (approved):
tACK f6a26196cfb7e2c90e25f82b0e2f569a05013cae

Thanks for picking this up and swiftly addressing feedback!

ItIsOHM commented:
> tACK [f6a2619](https://github.com/bitcoin/bitcoin/commit/f6a26196cfb7e2c90e25f82b0e2f569a05013cae)
> 
> Thanks for picking this up and swiftly addressing feedback!

Thanks so much to each and every reviewer here for tolerating a beginner contributor like me 😅

russeree commented:
Concept ACK

I do have a few questions about the implementation of the RPCHelpMan parameter text for 'longpollid' when used for a template instead of a propsal. My primary issue is the block proposal definition for ```longpollid``` at https://github.com/bitcoin/bitcoin/blob/57b8336dfed6312003cf34cd5ae7099f77115e73/src/rpc/mining.cpp#L613 
varies from the current PRs RPCHelpMan when used as an optional parameter for a block template. But the use of the passed parameter for ```longpollid``` at https://github.com/bitcoin/bitcoin/blob/57b8336dfed6312003cf34cd5ae7099f77115e73/src/rpc/mining.cpp#L715-L721 is the same for both a template and a proposal. 

**The questions I have are**

- Should the RPCHelpMan text for ```longpollid``` match for proposals and templates or does the behavior vary based on if the user is passing ```longpollid``` for a proposal vs template? 

- Should there be a mention of the 60 second timeout if we are going to describe not just the id component of the longpollid but the behavior of the optional parameter?

I apologize if I missed something completely.

stickies-v commented:
> But the use of the passed parameter for `longpollid` at
> https://github.com/bitcoin/bitcoin/blob/57b8336dfed6312003cf34cd5ae7099f77115e73/src/rpc/mining.cpp#L715-L721
> 
> is the same for both a template and a proposal.

`longpollid` is neither used as a parameter nor returned as a result for `mode==proposal`. The line you reference is never reached in that case, as we [return or throw within this code block](https://github.com/bitcoin/bitcoin/blob/57b8336dfed6312003cf34cd5ae7099f77115e73/src/rpc/mining.cpp#L660-L687). See also [the corresponding results](https://github.com/bitcoin/bitcoin/blob/57b8336dfed6312003cf34cd5ae7099f77115e73/src/rpc/mining.cpp#L573-L574).

> * Should the RPCHelpMan text for `longpollid` match for proposals and templates or does the behavior vary based on if the user is passing `longpollid` for a proposal vs template?

As before, it's not used for `proposal` mode. Could mention that in the description but I don't think it's that relevant, conceptually longpolling on a proposal doesn't make sense I think.

> * Should there be a mention of the 60 second timeout if we are going to describe not just the id component of the longpollid but the behavior of the optional parameter?

It only "times out" if there are new transactions to be considered [as per this `break`](https://github.com/bitcoin/bitcoin/blob/57b8336dfed6312003cf34cd5ae7099f77115e73/src/rpc/mining.cpp#L749), so I think the current description works well. Moreover, we may want to change this implementation detail (it's not part of the BIP22 spec), so keeping it out of the documentation makes that an easier process.

russeree commented:
> > 
> > is the same for both a template and a proposal.
> 
> `longpollid` is neither used as a parameter nor returned as a result for `mode==proposal`. The line you reference is never reached in that case, as we [return or throw within this code block](https://github.com/bitcoin/bitcoin/blob/57b8336dfed6312003cf34cd5ae7099f77115e73/src/rpc/mining.cpp#L660-L687). See also [the corresponding results](https://github.com/bitcoin/bitcoin/blob/57b8336dfed6312003cf34cd5ae7099f77115e73/src/rpc/mining.cpp#L573-L574).
> 
> > * Should the RPCHelpMan text for `longpollid` match for proposals and templates or does the behavior vary based on if the user is passing `longpollid` for a proposal vs template?
> 
> As before, it's not used for `proposal` mode. Could mention that in the description but I don't think it's that relevant, conceptually longpolling on a proposal doesn't make sense I think.

Thank you for this excellent overview, Thank you for taking the time so really dive into how this works. Sorry for being incorrect. 

> 
> > * Should there be a mention of the 60 second timeout if we are going to describe not just the id component of the longpollid but the behavior of the optional parameter?
> 
> It only "times out" if there are new transactions to be considered [as per this `break`](https://github.com/bitcoin/bitcoin/blob/57b8336dfed6312003cf34cd5ae7099f77115e73/src/rpc/mining.cpp#L749), so I think the current description works well. Moreover, we may want to change this implementation detail (it's not part of the BIP22 spec), so keeping it out of the documentation makes that an easier process.

Makes sense.

tACK https://github.com/bitcoin/bitcoin/commit/f6a26196cfb7e2c90e25f82b0e2f569a05013cae

luke-jr reviewed (approved):
utACK

luke-jr commented:
(Maybe remove # from PR title)
